File: algorithm/cpp/word-ladder.cpp
// 127. Word Ladder
// Difficulty: Medium

// Given two words (beginWord and endWord), and a dictionary word list,
// find the length of shortest transformation sequence from beginWord to endWord, such that:

// Only one letter can be changed at a time
// Each intermediate word must exist in the word list
// For example,

// Given:
// beginWord = "hit"
// endWord = "cog"
// wordList = ["hot","dot","dog","lot","log"]
// As one shortest transformation is "hit" -> "hot" -> "dot" -> "dog" -> "cog",
// return its length 5.

// Note:
// Return 0 if there is no such transformation sequence.
// All words have the same length.
// All words contain only lowercase alphabetic characters.


class Solution {
public:
    int ladderLength(string beginWord, string endWord, unordered_set<string>& wordDict) {
        unordered_map<string, int> pathLength;
        queue<string> q;
        pathLength[beginWord] = 1;
        q.push(beginWord);

        while (!q.empty()) {
            string word = q.front(); q.pop();
            for (int i = 0; i < word.size(); ++i) {
                string newWord = word;
                for (char ch = 'a'; ch <= 'z'; ++ch) {
                    newWord[i] = ch;
                    if (newWord == endWord) {
                    	return pathLength[word] + 1;
                    }
                    if (wordDict.count(newWord) && !pathLength.count(newWord)) {
                        q.push(newWord);
                        pathLength[newWord] = pathLength[word] + 1;
                    }   
                }
            }
        }
        return 0;
    }
};

File: data-structure-and-algorithm/typescript/leetcode/src/word-ladder.ts
/*
A transformation sequence from word beginWord to word endWord using a dictionary wordList is a sequence of words beginWord -> s1 -> s2 -> ... -> sk such that:

Every adjacent pair of words differs by a single letter.
Every si for 1 <= i <= k is in wordList. Note that beginWord does not need to be in wordList.
sk == endWord
Given two words, beginWord and endWord, and a dictionary wordList, return the number of words in the shortest transformation sequence from beginWord to endWord, or 0 if no such sequence exists.



Example 1:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log","cog"]
Output: 5
Explanation: One shortest transformation sequence is "hit" -> "hot" -> "dot" -> "dog" -> cog", which is 5 words long.
Example 2:

Input: beginWord = "hit", endWord = "cog", wordList = ["hot","dot","dog","lot","log"]
Output: 0
Explanation: The endWord "cog" is not in wordList, therefore there is no valid transformation sequence.


Constraints:

1 <= beginWord.length <= 10
endWord.length == beginWord.length
1 <= wordList.length <= 5000
wordList[i].length == beginWord.length
beginWord, endWord, and wordList[i] consist of lowercase English letters.
beginWord != endWord
All the words in wordList are unique.
*/

function ladderLength(begin, end, words) {
  const set = new Set([...words]);
  let q = [begin];
  let level = 1;
  for (; q.length; ++level) {
    const next = [];
    for (const curr of q) {
      if (curr === end) {
        return level;
      }
      set.delete(curr);
      for (let i = 0; i < curr.length; ++i) {
        const array = [...curr];
        for (let j = 0; j < 26; ++j) {
          array[i] = String.fromCharCode('a'.charCodeAt() + j);
          const word = array.join('');
          if (set.has(word)) {
            next.push(word);
          }
        }
      }
    }
    q = next;
  }
  return 0;
}
